File: dsa_leetcode_beginner/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return maxSum;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int leftGain = Math.max(dfs(node.left), 0);
        int rightGain = Math.max(dfs(node.right), 0);

        int priceNewPath = node.val + leftGain + rightGain;
        maxSum = Math.max(maxSum, priceNewPath);

        return node.val + Math.max(leftGain, rightGain);
    }
}

File: dsa_leetcode_beginner/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.js
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    constructor() {
        this.maxSum = -Infinity;
    }

    maxPathSum(root) {
        this.dfs(root);
        return this.maxSum;
    }

    dfs(node) {
        if (node === null) {
            return 0;
        }

        const leftGain = Math.max(this.dfs(node.left), 0);
        const rightGain = Math.max(this.dfs(node.right), 0);

        const priceNewPath = node.val + leftGain + rightGain;
        this.maxSum = Math.max(this.maxSum, priceNewPath);

        return node.val + Math.max(leftGain, rightGain);
    }
}
